File: src/common/web/utils/Strings.ts
/**
 * Converts an arbitrary value or an array of such to a string array.
 *
 * @param value - The value to convert.
 * @param delimiter - The delimiter to use for splitting a non-array value.
 */
export function convertToStringArray(value: any | any[], delimiter: string = " "): string[] {
    if (value === null || value === undefined) {
        return [];
    } else if (Array.isArray(value)) {
        let arr: string[] = [];
        value.forEach((obj) => {
            arr.push(String(obj));
        });
        return arr;
    } else {
        return String(value).split(delimiter);
    }
}

/**
 * Converts a file size (in bytes) to a human-readable form.
 *
 * @param size - The file size in bytes.
 *
 * @returns - A string representation of the size.
 */
export function humanReadableFileSize(size: number): string {
    try {
        const i = size == 0 ? 0 : Math.floor(Math.log(size) / Math.log(1024));
        return (size / Math.pow(1024, i)).toFixed(1) + " " + ["B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB"][i];
    } catch {
        return size + " B";
    }
}

/**
 * Formats a UNIX timestamp to a string using the system locale.
 *
 * @param date - The timestamp.
 *
 * @returns - The formatted string.
 */
export function formatLocaleTimestamp(date: number): string {
    return new Intl.DateTimeFormat(navigator.language, { dateStyle: "medium", timeStyle: "short" }).format(date * 1000);
}

/**
 * Adds a full stop to a string if necessary.
 *
 * @param sentence - The sentence to complete.
 */
export function finishSentence(sentence: string): string {
    if (sentence.length > 0 && sentence[sentence.length - 1] != ".") {
        sentence += ".";
    }
    return sentence;
}

/**
 * Formats a number of seconds into a readable form.
 *
 * @param elapsed - The elapsed time in seconds.
 *
 * @returns - The formatted string.
 */
export function formatElapsedTime(elapsed: number): string {
    const days = Math.floor(elapsed / (60 * 60 * 24));
    elapsed -= days * (60 * 60 * 24);

    const hours = Math.floor(elapsed / (60 * 60));
    elapsed -= hours * (60 * 60);

    const mins = Math.floor(elapsed / 60);
    elapsed -= mins * 60;

    const seconds = Math.floor(elapsed);
    elapsed -= seconds;

    const tokens: String[] = [];

    function adcToken(val: number, name: string) {
        if (val >= 1.0) {
            tokens.push(`${val}${name}`);
        }
    }

    adcToken(days, "d");
    adcToken(hours, "h");
    adcToken(mins, "m");
    adcToken(seconds, "s");

    return tokens.join(" ");
}

File: src/common/py/utils/strings.py
import typing


def human_readable_file_size(size: int, suffix="B") -> str:
    """
    Converts a file size to a human-readable string.
    
    Args:
        size: The file size to convert.
        suffix: Default suffix.

    Returns:
        The human-readable file size.
    """
    for unit in ("", "K", "M", "G", "T", "P", "E", "Z"):
        if abs(size) < 1024.0:
            return f"{size:3.2f} {unit}{suffix}"
        size /= 1024.0
    
    return f"{size:.2f} Y{suffix}"


def format_elapsed_time(elapsed: float) -> str:
    """
    Converts a number of seconds into a readable string.
    
    Args:
        elapsed: The elapsed time in seconds.

    Returns:
        The string representation.
    """
    m, s = divmod(elapsed, 60)
    h, m = divmod(m, 60)
    d, h = divmod(h, 60)
    
    tokens: typing.List[str] = []
    
    def add_token(val: float, name: str) -> None:
        if val >= 1.0:
            tokens.append(f"{val:.0f}{name}")
            
    add_token(d, "d")
    add_token(h, "h")
    add_token(m, "m")
    add_token(s, "s")
    
    return " ".join(tokens)


def ensure_starts_with(s: str, start: str) -> str:
    """
    Ensures that a string starts with a given string.
    
    Args:
        s: The string.
        start: The start to ensure.

    Returns:
        The new string.
    """
    if not s.startswith(start):
        s = start + s
    return s
